src/main.rs

```diff
@@ -151,48 +161,76 @@ fn elf_to_tbf<W: Write>(
 
     // Check for fixed addresses.
     //
-    // We do this by looking at segments in the ELF, and seeing if the virtual
-    // addresses match what expect for Tock PIC, or if they seem to be specific
-    // fixed addresses that the app was compiled for. This is complicated
-    // because there might be multiple segments for flash or RAM, even if the
-    // app is compiled for PIC.
+    // We do this with different mechanisms for flash and ram. For flash, we
+    // look at loadable segments in the ELF, and find the segment with the
+    // lowest address that is both executable and non-zero in file size. Since
+    // this is a segment that must be loaded to execute the application, we know
+    // that this must be a flash address.
     //
-    // We use the following algorithm:
-    // - If there is a segment that matches the PIC address we expect, then we
-    //   assume the app is PIC, and we do not include the matching fixed address
-    //   address.
-    // - Otherwise, we use the lowest address that we find as the fixed address.
+    // For RAM, we can't quite use the ELF file in the same way as with flash.
+    // Since nothing _actually_ has to be loaded into RAM, the ELF file does not
+    // have to keep track of the first address in RAM. Further, Tock apps
+    // typically put a .stack section at the beginning of RAM, and the stack is
+    // just a memory holder and doesn't contain any actual data. The start of
+    // RAM address will almost certainly exist somewhere in the ELF, but
+    // reliably extracting it from different ELFs linked with different
+    // toolchains from different linker scripts would I think require resorting
+    // to a bit of heuristics and guessing. To avoid the potential issues there,
+    // we instead require that a `_sram_origin` symbol be present to explicitly
+    // mark the start of RAM.
+    //
+    // In both cases we check to see if the address matches our expected PIC
+    // addresses:
+    // - RAM: 0x00000000
+    // - flash: 0x80000000
+    //
+    // These addresses are a Tock convention and enables PIC fixups to be done
+    // by the app when it first starts. If for some reason an app is PIC and
+    // wants to use different dummy PIC addresses, then this logic will have to
+    // be updated.
```

Comment:
Not for this PR, but I think this logic shouldn't change; rather these constant should just be exposed as flags.